template-coq/src/pluginCore.ml

```diff
@@ -1,57 +1,166 @@
 (* this file is the interface that extracted plugins will use.
  *)
 
-type ident   (* Template.Ast.ident *)
-type kername (* Template.Ast.kername *)
-type reductionStrategy (* Template.TemplateMonad.Common.reductionStrategy *)
-type global_reference (* Template.Ast.global_reference *)
+type ident   = Names.Id.t (* Template.Ast.ident *)
+type kername = Names.KerName.t (* Template.Ast.kername *)
+type reduction_strategy = Redexpr.red_expr (* Template.TemplateMonad.Common.reductionStrategy *)
+type global_reference = Globnames.global_reference (* Template.Ast.global_reference *)
 type term = Constr.t  (* Ast.term *)
+type mutual_inductive_body = Declarations.mutual_inductive_body (* Template.Ast.mutual_inductive_body *)
+type constant_entry = Declarations.constant_body (* Template.Ast.constant_entry *)
+type mutual_inductive_entry = Entries.mutual_inductive_entry (* Template.Ast.mutual_inductive_entry *)
 
-type 'a tm = 'a
+let default_flags = Redops.make_red_flag Genredexpr.[FBeta;FMatch;FFix;FCofix;FZeta;FDeltaBut []]
+let rs_cbv = Genredexpr.Cbv default_flags
+let rs_cbn = Genredexpr.Cbn default_flags
+let rs_hnf = Genredexpr.Hnf
+let rs_all = Genredexpr.Cbv Redops.all_flags
+let rs_lazy = Genredexpr.Cbv Redops.all_flags
 
-let not_implemented (s : string) : 'a tm =
-  assert false
+(* question(gmm): maybe this should just take a global_reference? *)
+let rs_unfold (env : Environ.env) (kn : kername) =
+  (* note(gmm): this shouldn't be necessary *)
+  let ident = Names.Id.of_string (Names.KerName.to_string kn) in
+  try
+    let gr = (Nametab.global (CAst.make (Libnames.Qualid (Libnames.qualid_of_ident ident)))) in
+    Genredexpr.Unfold [Locus.AllOccurrences,
+                       Tacred.evaluable_of_global_reference env gr]
+  with
+  | _ -> CErrors.user_err Pp.(str "Constant not found or not a constant: " ++ str (Names.Id.to_string ident))
 
-let tmReturn (x : 'a) : 'a tm = x
-let tmBind (x : 'a) (k : 'a -> 'b tm) : 'b tm =
-  k x
+type 'a tm =
+  { run_tm : Environ.env -> Evd.evar_map ->
+      (Environ.env -> Evd.evar_map -> 'a -> unit) -> (unit -> unit) -> unit }
+  (* note(gmm): i can't make this generic because of the way that
+   * tmLemma works. in particular, it needs to create a definition
+   * and then return, the result is passed through a hook.
+   *)
+
```

Comment:
Yep, no other choice than CPS here.

Reply:
Originally I made a record so that I could make it polymorphic, but it seems like that isn't really possible. Do you think I should drop the record and just convert this into an alias? I tend to like the added abstraction, but I don't think it is free in Ocaml the way it is with `newtype` in Haskell.

Reply:
I don't really have a preference, but indeed a record might incur a small performance penalty.